firedrake/projection.py

```diff
@@ -35,72 +81,129 @@ def project(v, V, bcs=None, mesh=None,
-        else:
-            deg = V.ufl_element().degree()
+class Assigner(object):
+    def __init__(self, source, target):
+        self.source = source
+        self.target = target
 
-        if v.rank() == 0:
-            fs = functionspace.FunctionSpace(V.mesh(), 'DG', deg+1)
-        elif v.rank() == 1:
-            fs = functionspace.VectorFunctionSpace(V.mesh(), 'DG',
-                                                   deg+1,
-                                                   dim=shape[0])
+    def project(self):
+        self.target.assign(self.source)
+        return self.target
+
+
+class ProjectorBase(object, metaclass=abc.ABCMeta):
+    def __init__(self, source, target, bcs=None, solver_parameters=None,
+                 form_compiler_parameters=None, constant_jacobian=True):
+        if solver_parameters is None:
+            solver_parameters = {}
         else:
-            fs = functionspace.TensorFunctionSpace(V.mesh(), 'DG',
-                                                   deg+1,
-                                                   shape=shape)
-        f = function.Function(fs)
-        f.interpolate(v)
-        v = f
-    elif isinstance(v, function.Function):
-        if v.function_space().mesh() != ret.function_space().mesh():
-            raise RuntimeError("Can't project between mismatching meshes")
-    elif not isinstance(v, ufl.core.expr.Expr):
-        raise RuntimeError("Can only project from expressions and functions, not %r" % type(v))
-
-    if v.ufl_shape != ret.ufl_shape:
-        raise RuntimeError('Shape mismatch between source %s and target function spaces %s in project' %
-                           (v.ufl_shape, ret.ufl_shape))
-
-    p = ufl_expr.TestFunction(V)
-    q = ufl_expr.TrialFunction(V)
-    a = ufl.inner(p, q) * ufl.dx(domain=V.mesh())
-    L = ufl.inner(p, v) * ufl.dx(domain=V.mesh())
-
-    # Default to 1e-8 relative tolerance
-    if solver_parameters is None:
-        solver_parameters = {'ksp_type': 'cg', 'ksp_rtol': 1e-8}
-    else:
-        solver_parameters.setdefault('ksp_type', 'cg')
-        solver_parameters.setdefault('ksp_rtol', 1e-8)
+            solver_parameters = solver_parameters.copy()
+        solver_parameters.setdefault("ksp_type", "cg")
+        self.source = source
+        self.target = target
+        self.solver_parameters = solver_parameters
+        self.form_compiler_parameters = form_compiler_parameters
+        self.bcs = bcs
+        self.constant_jacobian = constant_jacobian
+
+    @cached_property
+    def is_DG(self):
+        element = self.target.function_space().finat_element
+        return element.entity_closure_dofs() == element.entity_dofs()
+
+    @cached_property
+    def A(self):
+        u = firedrake.TrialFunction(self.target.function_space())
+        v = firedrake.TestFunction(self.target.function_space())
+        a = firedrake.inner(u, v)*firedrake.dx
+        if self.is_DG and self.bcs is None:
+            a = firedrake.Tensor(a).inv
```

Comment:
This is maybe bad, because for high order spaces (especially tensor-valued ones), we overflow Eigen's stack limits. Could remove it.

Reply:
I would also like to fix #1238. One possible option is to use dynamic matrix types where you no longer need to rely on stack limits and instead, these larger matrices are heap allocated. I raised an issue that @wence-  described in #1221. This is still a topic that needs further discussion. The compiler for Slate needs to detect when heap allocation is necessary (i.e. if the matrix is large enough to surpass Eigen's stack limit, we switch to dynamic types). I'm open to suggestions for what the best way to do this is.